src/analyzer/analyzersilence.cpp

```diff
@@ -43,34 +41,64 @@ bool AnalyzerSilence::initialize(TrackPointer pTrack,
     }
 
     m_iFramesProcessed = 0;
-    m_bPrevSilence = true;
     m_iSignalStart = -1;
     m_iSignalEnd = -1;
 
     return true;
 }
 
-bool AnalyzerSilence::processSamples(const CSAMPLE* pIn, const int iLen) {
-    for (int i = 0; i < iLen; i += mixxx::kAnalysisChannels) {
-        // Compute max of channels in this sample frame
-        CSAMPLE fMax = CSAMPLE_ZERO;
-        for (SINT ch = 0; ch < mixxx::kAnalysisChannels; ++ch) {
-            CSAMPLE fAbs = fabs(pIn[i + ch]);
-            fMax = math_max(fMax, fAbs);
+// static
+SINT AnalyzerSilence::findFirstSound(const CSAMPLE* pIn, SINT iLen) {
+    for (SINT i = 0; i < iLen; ++i) {
+        if (fabs(pIn[i]) >= kSilenceThreshold) {
+            return i;
         }
+    }
+    return -1;
+}
 
-        bool bSilence = fMax < m_fThreshold;
-
-        if (m_bPrevSilence && !bSilence) {
-            if (m_iSignalStart < 0) {
-                m_iSignalStart = m_iFramesProcessed + i / mixxx::kAnalysisChannels;
-            }
-        } else if (!m_bPrevSilence && bSilence) {
-            m_iSignalEnd = m_iFramesProcessed + i / mixxx::kAnalysisChannels;
+// static
+SINT AnalyzerSilence::findLastSound(const CSAMPLE* pIn, SINT iLen, SINT firstSound) {
+    DEBUG_ASSERT(firstSound >= -1);
+    SINT lastSound = firstSound;
+    for (SINT i = firstSound + 1; i < iLen; ++i) {
+        if (fabs(pIn[i]) >= kSilenceThreshold) {
+            lastSound = i;
         }
+    }
+    return lastSound;
+}
 
-        m_bPrevSilence = bSilence;
+// static
+bool AnalyzerSilence::verifyFirstSound(
+        const CSAMPLE* pIn, SINT iLen, mixxx::audio::FramePos firstSoundFrame) {
+    if (mixxx::audio::FramePos(findFirstSound(pIn, iLen) /
+                mixxx::kAnalysisChannels) == firstSoundFrame) {
+        qDebug() << "First sound found at the previously stored position";
+        return true;
     }
+
+    // This can happen in case of track edits or replacements, changed encoders or encoding issues.
+    qWarning() << "First sound has been moved! The beatgrid and "
```

Comment:
`AnalyzerSilence` cannot decide why this might be a problem in a different context. It just knows about some threshold and how to find the first and last sound in a given buffer depending on the threshold. That's it.

Reply:
Yes sure, but this is just a static message that gives the user more context in case of hitting such issue.
Once we have a GUI representation we can move writing the log entry elsewhere.